File: app/src/main/java/net/bloople/manga/BooksAdapter.java
package net.bloople.manga;

import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.PopupWindow;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import net.bloople.manga.audit.AuditEventsActivity;

import java.util.ArrayList;

class BooksAdapter extends RecyclerView.Adapter<BooksAdapter.ViewHolder> {
    private ArrayList<Book> books = new ArrayList<>();
    private ArrayList<Long> selectedBookIds = new ArrayList<>();
    private boolean selectable = false;

    BooksAdapter() {
        setHasStableIds(true);
    }

    public long getItemId(int position) {
        return books.get(position).getId();
    }

    public boolean isSelectable() {
        return selectable;
    }

    void setSelectable(boolean isSelectable) {
        selectable = isSelectable;
        notifyDataSetChanged();
    }

    ArrayList<Long> getSelectedBookIds() {
        return selectedBookIds;
    }

    void setSelectedBookIds(ArrayList<Long> selectedBookIds) {
        this.selectedBookIds = selectedBookIds;
        notifyDataSetChanged();
    }

    void clearSelectedBookIds() {
        setSelectedBookIds(new ArrayList<>());
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return books.size();
    }

    void update(ArrayList<Book> inBooks) {
        books = inBooks;
        notifyDataSetChanged();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        TextView pageCountView;
        TextView textView;
        ImageView imageView;
        ImageView selectableView;
        ViewHolder(View view) {
            super(view);

            pageCountView = view.findViewById(R.id.page_count_view);
            textView = view.findViewById(R.id.text_view);
            imageView = view.findViewById(R.id.image_view);
            selectableView = view.findViewById(R.id.selectable);

            view.setOnClickListener(v -> {
                Book book = books.get(getAdapterPosition());
                long bookId = book.getId();

                if(selectable) {
                    if(selectedBookIds.contains(bookId)) {
                        selectedBookIds.remove(bookId);
                        v.setActivated(false);
                    }
                    else {
                        selectedBookIds.add(bookId);
                        v.setActivated(true);
                    }

                    notifyItemChanged(getAdapterPosition());
                }
                else {
                    openBook(book, true);
                }
            });

            view.setOnLongClickListener(v -> {
                if(selectable) return false;

                Book book = books.get(getAdapterPosition());
                openBook(book, false);

                return true;
            });

            textView.setOnClickListener(v -> {
                Book book = books.get(getAdapterPosition());
                showFullBookTitle(book);
            });

            textView.setOnLongClickListener(v -> {
                Book book = books.get(getAdapterPosition());
                showBookTags(book);

                return true;
            });
        }

        private void openBook(Book book, boolean resume) {
            Intent intent = new Intent(itemView.getContext(), ReadingActivity.class);
            intent.putExtra("_id", book.getId());
            intent.putExtra("resume", resume);
            intent.putExtra("libraryId", book.getLibrary().id());

            itemView.getContext().startActivity(intent);
        }

        private void showFullBookTitle(Book book) {
            BookMetadata metadata = BookMetadata.findOrCreateByBookId(itemView.getContext(), book.getId());

            View popupView = LayoutInflater.from(itemView.getContext()).inflate(R.layout.index_book_title_popup, null, false);
            TextView bookTitleView = popupView.findViewById(R.id.book_title);
            bookTitleView.setText(book.getTitle() + "\nOpened Count: " + metadata.openedCount());

            ImageButton viewAuditEventsButton = popupView.findViewById(R.id.view_audit_events);
            viewAuditEventsButton.setOnClickListener(v -> {
                Intent intent = new Intent(viewAuditEventsButton.getContext(), AuditEventsActivity.class);
                intent.putExtra("resourceId", book.getId());

                viewAuditEventsButton.getContext().startActivity(intent);
            });

            int popupWidth = textView.getWidth() + bookTitleView.getPaddingStart() + bookTitleView.getPaddingEnd();
            PopupWindow popup = new PopupWindow(popupView, popupWidth, ViewGroup.LayoutParams.WRAP_CONTENT);
            popup.setFocusable(true);
            popup.setOutsideTouchable(true);
            popup.setElevation(24);

            popupView.setOnClickListener(v -> {
                popup.dismiss();
            });

            popup.showAsDropDown(textView, -bookTitleView.getPaddingStart(), -textView.getHeight(), Gravity.TOP | Gravity.START);
        }

        private void showBookTags(Book book) {
            IndexActivity indexActivity = (IndexActivity)itemView.getContext();

            TagChooserFragment tagChooser = TagChooserFragment.newInstance(book.getTags().toArray(new String[0]));
            tagChooser.show(indexActivity.getSupportFragmentManager(), "tag_chooser");
        }
    }

    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public BooksAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.index_book_view, parent, false);

        final double viewWidthToBitmapWidthRatio = (double)parent.getWidth() / 4.0 / 197.0;
        view.getLayoutParams().height = (int)(310.0 * viewWidthToBitmapWidthRatio);

        return new ViewHolder(view);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Book book = books.get(position);

        holder.selectableView.setVisibility(selectable ? View.VISIBLE : View.INVISIBLE);
        if(selectable) holder.itemView.setActivated(selectedBookIds.contains(book.getId()));

        //holder.textView.setText(title.substring(0, Math.min(50, title.length())));
        holder.textView.setText(book.getTitle());

        holder.pageCountView.setText(String.format("%,d", book.getPages()));

        Glide.clear(holder.imageView);
        holder.imageView.setImageDrawable(null);

        Glide.with(holder.imageView.getContext()).load(book.thumbnailUrl().toGlideUrl()).dontAnimate().into(holder.imageView);
    }

}

File: app/src/main/java/net/bloople/manga/BooksAdapter.kt
package net.bloople.manga

import com.bumptech.glide.RequestManager
import com.bumptech.glide.util.ViewPreloadSizeProvider
import com.bumptech.glide.load.model.GlideUrl
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.ListPreloader.PreloadModelProvider
import android.widget.TextView
import android.content.Intent
import android.view.LayoutInflater
import android.widget.ImageButton
import net.bloople.manga.audit.AuditEventsActivity
import android.widget.PopupWindow
import android.view.ViewGroup
import android.view.Gravity
import com.bumptech.glide.RequestBuilder
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.ImageView
import java.util.ArrayList

internal class BooksAdapter(requestManager: RequestManager, preloadSizeProvider: ViewPreloadSizeProvider<GlideUrl>) :
    RecyclerView.Adapter<BooksAdapter.ViewHolder>(), PreloadModelProvider<GlideUrl> {
    private val requestManager: RequestManager
    private val preloadSizeProvider: ViewPreloadSizeProvider<GlideUrl>
    private var books = ArrayList<Book>()
    private var booksMetadata = HashMap<Long, BookMetadata>();

    init {
        setHasStableIds(true)
        this.requestManager = requestManager
        this.preloadSizeProvider = preloadSizeProvider
    }

    override fun getItemId(position: Int): Long {
        return books[position].id
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return books.size
    }

    fun update(searchResults: SearchResults) {
        books = searchResults.books
        booksMetadata = searchResults.booksMetadata
        notifyDataSetChanged()
    }

    internal inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var pageCountView: TextView = view.findViewById(R.id.page_count_view)
        var openedCountView: TextView = view.findViewById(R.id.opened_count_view)
        var textView: TextView = view.findViewById(R.id.text_view)
        var imageView: ImageView = view.findViewById(R.id.image_view)

        init {
            view.setOnClickListener { v: View ->
                val book = books[bindingAdapterPosition]
                openBook(book, true)
            }

            view.setOnLongClickListener {
                val book = books[bindingAdapterPosition]
                openBook(book, false)

                true
            }

            textView.setOnClickListener {
                val book = books[bindingAdapterPosition]
                showFullBookTitle(book)
            }

            textView.setOnLongClickListener {
                val book = books[bindingAdapterPosition]
                showBookTags(book)

                true
            }
        }

        private fun openBook(book: Book, resume: Boolean) {
            val intent = Intent(itemView.context, ReadingActivity::class.java)
            intent.putExtra("_id", book.id)
            intent.putExtra("resume", resume)
            intent.putExtra("libraryId", book.library.id)

            itemView.context.startActivity(intent)
        }

        private fun showFullBookTitle(book: Book) {
            val popupView = LayoutInflater.from(itemView.context).inflate(R.layout.index_book_title_popup, null, false)
            val bookTitleView: TextView = popupView.findViewById(R.id.book_title)
            bookTitleView.text = book.title

            val viewAuditEventsButton: ImageButton = popupView.findViewById(R.id.view_audit_events)
            viewAuditEventsButton.setOnClickListener {
                val intent = Intent(viewAuditEventsButton.context, AuditEventsActivity::class.java)
                intent.putExtra("resourceId", book.id)

                viewAuditEventsButton.context.startActivity(intent)
            }

            val popupWidth = textView.width + bookTitleView.paddingStart + bookTitleView.paddingEnd
            val popup = PopupWindow(popupView, popupWidth, ViewGroup.LayoutParams.WRAP_CONTENT)
            popup.isFocusable = true
            popup.isOutsideTouchable = true
            popup.elevation = 24f

            popupView.setOnClickListener { popup.dismiss() }

            popup.showAsDropDown(textView, -bookTitleView.paddingStart, -textView.height, Gravity.TOP or Gravity.START)
        }

        private fun showBookTags(book: Book) {
            val indexActivity = itemView.context as IndexActivity

            val tagChooser = TagChooserFragment.newInstance(book.tags.toTypedArray())
            tagChooser.show(indexActivity.supportFragmentManager, "tag_chooser")
        }
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.index_book_view, parent, false)

        val viewWidthToBitmapWidthRatio = parent.width.toDouble() / 4.0 / 197.0
        view.layoutParams.height = (310.0 * viewWidthToBitmapWidthRatio).toInt()

        preloadSizeProvider.setView(view)

        return ViewHolder(view)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val book = books[position]
        val metadata = booksMetadata[book.id]

        //holder.textView.setText(title.substring(0, Math.min(50, title.length())));
        holder.textView.text = book.title

        holder.pageCountView.text = String.format("%,d", book.pages)

        val openedCount = metadata?.openedCount ?: 0
        if(openedCount > 0) {
            holder.openedCountView.visibility = View.VISIBLE
            holder.openedCountView.text = String.format("%,d", openedCount)
        }
        else {
            holder.openedCountView.visibility = View.GONE
        }

        requestManager
            .load(book.thumbnailUrl.toGlideUrl())
            .into(holder.imageView)
    }

    override fun getPreloadItems(position: Int): List<GlideUrl> {
        val book = books[position]
        return listOf(book.thumbnailUrl.toGlideUrl())
    }

    override fun getPreloadRequestBuilder(url: GlideUrl): RequestBuilder<Drawable> {
        return requestManager.load(url)
    }
}